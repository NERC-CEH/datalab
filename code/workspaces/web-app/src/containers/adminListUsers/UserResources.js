import React from 'react';
import { useDispatch } from 'react-redux';
import Checkbox from '@material-ui/core/Checkbox';
import Select from '@material-ui/core/Select';
import MenuItem from '@material-ui/core/MenuItem';
import Typography from '@material-ui/core/Typography';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import { makeStyles } from '@material-ui/core/styles';
import { permissionTypes } from 'common';
import { useCurrentUserId } from '../../hooks/authHooks';
import { ResourceAccordion, ResourceAccordionSummary, ResourceAccordionDetails } from '../../components/common/ResourceAccordion';
import userSummary from './userSummary';
import projectsToShow from './projectsToShow';
import UserProject from './UserProject';
import Pagination from '../../components/stacks/Pagination';
import roleActions from '../../actions/roleActions';
import { getCatalogue } from '../../config/catalogue';

const { CATALOGUE_ROLES } = permissionTypes;

const useStyles = makeStyles(theme => ({
  container: {
    marginBottom: theme.spacing(10),
  },
  resources: {
    marginLeft: theme.spacing(8),
  },
  heading: {
    margin: [[theme.spacing(2), 0]],
  },
  summary: {
    display: 'flex',
    flexDirection: 'column',
  },
  systemRoles: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
  },
  systemRoleSkeletonItem: {
    marginRight: theme.spacing(2),
  },
  placeHolderCard: {
    width: '100%',
    height: 70,
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    borderTop: `1px solid ${theme.palette.divider}`,
    borderBottom: `1px solid ${theme.palette.divider}`,
    marginTop: theme.spacing(2),
  },
  systemSelect: {
    marginLeft: theme.spacing(2),
  },
}));

export default function UserResources({ user, filters, roles }) {
  const dispatch = useDispatch();
  const classes = useStyles();
  const currentUserId = useCurrentUserId();
  const catalogue = getCatalogue();

  if (!roles) {
    return null;
  }
  const projects = projectsToShow(filters, roles);

  const renderedProjects = projects && projects.length > 0
    ? projects.map(projectKey => <UserProject key={projectKey} userId={user.userId} projectKey={projectKey} filters={filters} roles={roles} />)
    : [<div className={classes.placeHolderCard} key={'place-holder-card'}>
      <Typography variant="body1">No projects to display.</Typography>
    </div>];

  const handleSystemCheckbox = (event) => {
    switch (event.target.name) {
      case 'instanceAdmin':
        dispatch(roleActions.setInstanceAdmin(user.userId, event.target.checked));
        break;
      case 'dataManager':
        dispatch(roleActions.setDataManager(user.userId, event.target.checked));
        break;
      default:
        // no default action
    }
  };

  const handleSystemSelect = (event) => {
    dispatch(roleActions.setCatalogueRole(user.userId, event.target.value));
  };

  const SystemCheckbox = ({ label, checked, name, disabled }) => {
    const id = `system-checkbox-${user.userId}-${name}`;
    return (
      <>
        <Typography id={id} variant="body1">{label}</Typography>
        <Checkbox checked={checked} onChange={handleSystemCheckbox} name={name} color="primary" disabled={disabled} aria-labelledby={id} />
      </>
    );
  };

  const SystemSelect = ({ itemPrefix, current, items, name }) => (
    <div className={classes.systemSelect}>
      <Select value={current} onChange={handleSystemSelect} name={name} color="primary" variant="outlined" margin="dense" aria-label={`${itemPrefix} role`}>
        {items.map(item => (
          <MenuItem key={item} value={item}>{itemPrefix} {item}</MenuItem>
        ))}
      </Select>
    </div>
  );

  return (
    <div className={classes.container}>
      <ResourceAccordion>
        <ResourceAccordionSummary expandIcon={<ExpandMoreIcon />}>
          <div className={classes.summary}>
            <Typography variant="h5" className={classes.heading}>{user.name}</Typography>
            <Typography variant="body2">{userSummary(filters, roles)}</Typography>
          </div>
        </ResourceAccordionSummary>
        <ResourceAccordionDetails>
          <div className={classes.resources}>
            <div className={classes.systemRoles}>
              <SystemCheckbox label="Instance admin" checked={roles.instanceAdmin} name="instanceAdmin" disabled={user.userId === currentUserId} />
              <SystemCheckbox label="Data manager" checked={roles.dataManager} name="dataManager" />
              {catalogue.available && <SystemSelect itemPrefix="Catalogue" current={roles.catalogueRole} items={CATALOGUE_ROLES} name="catalogue" />}
            </div>
            <Pagination items={renderedProjects} itemsPerPage={5} itemsName="Projects" />
          </div>
        </ResourceAccordionDetails>
      </ResourceAccordion>
    </div>
  );
}
