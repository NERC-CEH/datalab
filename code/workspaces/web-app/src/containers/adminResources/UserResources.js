import React from 'react';
import Checkbox from '@material-ui/core/Checkbox';
import Typography from '@material-ui/core/Typography';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import { makeStyles } from '@material-ui/core/styles';
import { ResourceAccordion, ResourceAccordionSummary, ResourceAccordionDetails } from './ResourceAccordion';
import allFiltersOff from './allFiltersOff';

const useStyles = makeStyles(theme => ({
  container: {
    marginBottom: theme.spacing(10),
  },
  resources: {
    marginLeft: theme.spacing(8),
  },
  heading: {
    margin: [[theme.spacing(2), 0]],
  },
  summary: {
    display: 'flex',
    flexDirection: 'column',
  },
  instanceAdmin: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'baseline',
  },
}));

export function userSummary(filters, roles) {
  const filtersOff = allFiltersOff(filters);
  const info = [];
  const plural = num => (num !== 1 ? 's' : '');

  // project roles
  if (filtersOff || filters.projectViewer) {
    const numProjects = roles.projectViewer.length;
    info.push(`${numProjects} project${plural(numProjects)}`);
  } else if (filters.projectUser) {
    const numProjects = roles.projectUser.length;
    info.push(`${numProjects} project${plural(numProjects)}`);
  } else if (filters.projectAdmin) {
    const numProjects = roles.projectAdmin.length;
    info.push(`${numProjects} project${plural(numProjects)}`);
  }

  // site roles
  if (filtersOff || filters.siteOwner) {
    const numSites = roles.siteOwner.length;
    info.push(`${numSites} site${plural(numSites)}`);
  }

  // notebook roles
  if (filtersOff || filters.notebookOwner) {
    const numNotebooks = roles.notebookOwner.length;
    info.push(`${numNotebooks} notebook${plural(numNotebooks)}`);
  }

  // storage roles
  if (filtersOff || filters.storageAccess) {
    const numDataStores = roles.storageAccess.length;
    info.push(`${numDataStores} data store${plural(numDataStores)}`);
  }

  return info.join(', ');
}

export function projectsToShow(filters, roles) {
  const filtersOff = allFiltersOff(filters);
  const projectKeys = [];

  // project roles
  if (filtersOff || filters.projectViewer) {
    projectKeys.push(...roles.projectViewer);
  }
  if (filtersOff || filters.projectUser) {
    projectKeys.push(...roles.projectUser);
  }
  if (filtersOff || filters.projectAdmin) {
    projectKeys.push(...roles.projectAdmin);
  }

  // site roles
  if (filtersOff || filters.siteOwner) {
    projectKeys.push(...roles.siteOwner.map(resource => resource.projectKey));
  }

  // notebook roles
  if (filtersOff || filters.notebookOwner) {
    projectKeys.push(...roles.notebookOwner.map(resource => resource.projectKey));
  }

  // storage roles
  if (filtersOff || filters.storageAccess) {
    projectKeys.push(...roles.storageAccess.map(resource => resource.projectKey));
  }

  const uniqueProjectKeys = [...new Set(projectKeys)].sort();
  return uniqueProjectKeys;
}

export default function UserResources(props) {
  const classes = useStyles();
  const { userPermissions, user, filters, roles } = props;

  if (!roles) { return null; }
  const projects = projectsToShow(filters, roles);

  return (
    <div className={classes.container}>
      <ResourceAccordion key={user.userId} defaultExpanded>
        <ResourceAccordionSummary expandIcon={<ExpandMoreIcon />}>
          <div className={classes.summary}>
            <Typography variant="h5" className={classes.heading}>{user.name}</Typography>
            <Typography variant="body2">{userSummary(filters, roles)}</Typography>
          </div>
        </ResourceAccordionSummary>
        <ResourceAccordionDetails>
          <div className={classes.resources}>
            <div className={classes.instanceAdmin}>
              <Typography variant="h5">Instance admin</Typography>
              <Checkbox checked={roles.instanceAdmin} name="instanceAdmin" color="primary" disabled />
            </div>
            {JSON.stringify(projects)}
          </div>
        </ResourceAccordionDetails>
      </ResourceAccordion>
    </div>
  );
}
