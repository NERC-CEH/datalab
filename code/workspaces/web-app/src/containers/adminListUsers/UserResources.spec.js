import React from 'react';
import { shallow } from 'enzyme';
import UserResources from './UserResources';

const user = { name: 'user name' };
const projectKey = 'proj-1234';
const siteName = 'site name';
const notebookName = 'notebook name';
const storageName = 'storage name';

describe('UserResources', () => {
  const shallowRender = () => {
    const filters = {
      instanceAdmin: false,
      projectAdmin: false,
      projectUser: false,
      projectViewer: false,
      siteOwner: false,
      notebookOwner: false,
      storageAccess: false,
    };
    const roles = {
      instanceAdmin: true,
      projectAdmin: [projectKey],
      projectUser: [projectKey],
      projectViewer: [projectKey],
      siteOwner: [{ projectKey, name: siteName }],
      notebookOwner: [{ projectKey, name: notebookName }],
      storageAccess: [{ projectKey, name: storageName }],
    };
    const props = {
      user, filters, roles,
    };

    return shallow(<UserResources {...props} />);
  };

  it('renders to match snapshot passing correct props to children', () => {
    expect(shallowRender()).toMatchSnapshot();
  });
});
