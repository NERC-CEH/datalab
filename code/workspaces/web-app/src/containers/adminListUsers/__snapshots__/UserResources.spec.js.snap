// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserResources renders to match snapshot passing correct props to children when catalogue is available 1`] = `
<div>
  <div
    class="makeStyles-container-1"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root ForwardRef(Accordion)-root-9 MuiAccordion-rounded MuiAccordion-gutters css-16rgl2j-MuiPaper-root-MuiAccordion-root"
    >
      <div
        aria-expanded="false"
        class="MuiButtonBase-root MuiAccordionSummary-root ForwardRef(AccordionSummary)-root-11 MuiAccordionSummary-gutters css-ezupba-MuiButtonBase-root-MuiAccordionSummary-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiAccordionSummary-content ForwardRef(AccordionSummary)-content-14 MuiAccordionSummary-contentGutters css-o4b71y-MuiAccordionSummary-content"
        >
          <div
            class="makeStyles-summary-4"
          >
            <h5
              class="MuiTypography-root MuiTypography-h5 makeStyles-heading-3 css-k8n68h-MuiTypography-root"
            >
              user name
            </h5>
            <p
              class="MuiTypography-root MuiTypography-body2 css-1o1pkbg-MuiTypography-root"
            >
              1 project, 1 site, 1 notebook, 1 data store
            </p>
          </div>
        </div>
        <div
          class="MuiAccordionSummary-expandIconWrapper css-yw020d-MuiAccordionSummary-expandIconWrapper"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ExpandMoreIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </div>
      </div>
      <div
        class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
        style="min-height: 0px;"
      >
        <div
          class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
          >
            <div
              class="MuiAccordion-region"
              role="region"
            >
              <div
                class="MuiAccordionDetails-root ForwardRef(AccordionDetails)-root-15 css-1l0wl9e-MuiAccordionDetails-root"
              >
                <div
                  class="makeStyles-resources-2"
                >
                  <div
                    class="makeStyles-systemRoles-5"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-12gb3o9-MuiTypography-root"
                      id="system-checkbox-uid1-instanceAdmin"
                    >
                      Instance admin
                    </p>
                    <span
                      aria-labelledby="system-checkbox-uid1-instanceAdmin"
                      class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary Mui-checked MuiCheckbox-root MuiCheckbox-colorPrimary css-yql927-MuiButtonBase-root-MuiCheckbox-root"
                    >
                      <input
                        checked=""
                        class="PrivateSwitchBase-input css-1m9pwf3"
                        data-indeterminate="false"
                        name="instanceAdmin"
                        type="checkbox"
                      />
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="CheckBoxIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                        />
                      </svg>
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </span>
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-12gb3o9-MuiTypography-root"
                      id="system-checkbox-uid1-dataManager"
                    >
                      Data manager
                    </p>
                    <span
                      aria-labelledby="system-checkbox-uid1-dataManager"
                      class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary Mui-checked MuiCheckbox-root MuiCheckbox-colorPrimary css-yql927-MuiButtonBase-root-MuiCheckbox-root"
                    >
                      <input
                        checked=""
                        class="PrivateSwitchBase-input css-1m9pwf3"
                        data-indeterminate="false"
                        name="dataManager"
                        type="checkbox"
                      />
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="CheckBoxIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                        />
                      </svg>
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </span>
                    <div
                      class="makeStyles-systemSelect-8"
                    >
                      <div
                        aria-label="Catalogue role"
                        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary  css-di4u0j-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
                      >
                        <div
                          aria-expanded="false"
                          aria-haspopup="listbox"
                          aria-labelledby="mui-component-select-catalogue"
                          class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-1f6u0wk-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                          id="mui-component-select-catalogue"
                          role="button"
                          tabindex="0"
                        >
                          <span
                            class="notranslate"
                          >
                            ​
                          </span>
                        </div>
                        <input
                          aria-hidden="true"
                          class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                          name="catalogue"
                          tabindex="-1"
                          value=""
                        />
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                          data-testid="ArrowDropDownIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M7 10l5 5 5-5z"
                          />
                        </svg>
                        <fieldset
                          aria-hidden="true"
                          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                        >
                          <legend
                            class="css-ihdtdm"
                          >
                            <span
                              class="notranslate"
                            >
                              ​
                            </span>
                          </legend>
                        </fieldset>
                      </div>
                    </div>
                  </div>
                  <div>
                    <div>
                      user project
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`UserResources renders to match snapshot passing correct props to children when catalogue is not available 1`] = `
<div>
  <div
    class="makeStyles-container-17"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root ForwardRef(Accordion)-root-25 MuiAccordion-rounded MuiAccordion-gutters css-16rgl2j-MuiPaper-root-MuiAccordion-root"
    >
      <div
        aria-expanded="false"
        class="MuiButtonBase-root MuiAccordionSummary-root ForwardRef(AccordionSummary)-root-27 MuiAccordionSummary-gutters css-ezupba-MuiButtonBase-root-MuiAccordionSummary-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiAccordionSummary-content ForwardRef(AccordionSummary)-content-30 MuiAccordionSummary-contentGutters css-o4b71y-MuiAccordionSummary-content"
        >
          <div
            class="makeStyles-summary-20"
          >
            <h5
              class="MuiTypography-root MuiTypography-h5 makeStyles-heading-19 css-k8n68h-MuiTypography-root"
            >
              user name
            </h5>
            <p
              class="MuiTypography-root MuiTypography-body2 css-1o1pkbg-MuiTypography-root"
            >
              1 project, 1 site, 1 notebook, 1 data store
            </p>
          </div>
        </div>
        <div
          class="MuiAccordionSummary-expandIconWrapper css-yw020d-MuiAccordionSummary-expandIconWrapper"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ExpandMoreIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </div>
      </div>
      <div
        class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
        style="min-height: 0px;"
      >
        <div
          class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
          >
            <div
              class="MuiAccordion-region"
              role="region"
            >
              <div
                class="MuiAccordionDetails-root ForwardRef(AccordionDetails)-root-31 css-1l0wl9e-MuiAccordionDetails-root"
              >
                <div
                  class="makeStyles-resources-18"
                >
                  <div
                    class="makeStyles-systemRoles-21"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-12gb3o9-MuiTypography-root"
                      id="system-checkbox-uid1-instanceAdmin"
                    >
                      Instance admin
                    </p>
                    <span
                      aria-labelledby="system-checkbox-uid1-instanceAdmin"
                      class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary Mui-checked MuiCheckbox-root MuiCheckbox-colorPrimary css-yql927-MuiButtonBase-root-MuiCheckbox-root"
                    >
                      <input
                        checked=""
                        class="PrivateSwitchBase-input css-1m9pwf3"
                        data-indeterminate="false"
                        name="instanceAdmin"
                        type="checkbox"
                      />
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="CheckBoxIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                        />
                      </svg>
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </span>
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-12gb3o9-MuiTypography-root"
                      id="system-checkbox-uid1-dataManager"
                    >
                      Data manager
                    </p>
                    <span
                      aria-labelledby="system-checkbox-uid1-dataManager"
                      class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary Mui-checked MuiCheckbox-root MuiCheckbox-colorPrimary css-yql927-MuiButtonBase-root-MuiCheckbox-root"
                    >
                      <input
                        checked=""
                        class="PrivateSwitchBase-input css-1m9pwf3"
                        data-indeterminate="false"
                        name="dataManager"
                        type="checkbox"
                      />
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="CheckBoxIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                        />
                      </svg>
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </span>
                  </div>
                  <div>
                    <div>
                      user project
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
